### PR TITLE
Addition of announcement/notification banner

### PR DESCRIPTION
**Pull request by AndyKilmory**
## What does this change do?
This feature adds in a new React component to provide a notification/announcement banner that can display a variety of messages to the user. Initial use case will be to announce the release of new features but it has been designed so that in time it can be used for all user notifications - warnings, error messages, information etc.

The new features are announced by adding items to the announcements array in the application.conf

The addition of patterns to introduce other messages/notifications has been left for a future PR.

If no announcements are included in the application.conf the functionality will have no impact on the application.

The application polls the server one every minute to check for new notifications

The format of the announcement objects is;

[ (array)
    { (json object)
      **announceId**: (string) the unique id of the announcement - should be unique among all active announcements
      **description**: (string) the main text to display in the notification notification_banner
      **endDate**: (string, optional, format="yyyy-mm-dd") the date beyond which the announcement should not be seen, if not present set as today + 1 year
      **url**: (string, optional) a link to a page/document providing further details regarding announcement
      **urlText**: (string, optional) text to be included in a-tag hyperlink (will revert to default if not present)
      **category**: (string) the type of announcement - will control styling and display, Enum=_announcement, information, warning, error, success_
      **lifespan**: (string) the lifecycle behaviour Enum=_transient_ (message disappears on any click etc),
                                                      _session_ (message must be acknowledged but action NOT stored in client cookie - used for current session messages)
                                                      _persistent_ (message must be acknowledged and action stored in client cookie - used for long running announcements)
   },
   ...
 ]

## How should a reviewer test this change?

Add some example notifications to the application.conf and ensure that they displaty correctly and follow the correct lifetime rules (see lifespan options above)

A set of examples could be;

announcements = [
  {
    announceId: "notification_banner",
    description: "A New Feature Notification Banner",
    endDate: "2025-03-31",
    url: "https://runacharainn.com",
    urlText: "Runach Arainn Glamping",
    category: "announcement",
    lifespan: "persistent"
  },
  {
    announceId: "sort_control",
    description: "Sort Control",
    endDate: "2025-04-01",
    url: "https://www.bbc.co.uk",
    category: "warning",
    lifespan: "transient"
  },
  {
      announceId: "success_ex",
      description: "This is an example of a success message",
      endDate: "2025-04-01",
      category: "success",
      lifespan: "transient"
  },
  {
      announceId: "error_msg",
      description: "This is an example of a error message. Please click to acknowledge.",
      endDate: "2025-04-01",
      category: "error",
      lifespan: "session"
  },
  {
      announceId: "old_message",
      description: "This is an old message",
      endDate: "2024-01-01",
      category: "success",
      lifespan: "persistent"
  }
]

An example of how these notifications should look is;

Screenshot 2024-03-19 at 09 40 38

## Tested? Documented?
- [ ] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment
- [ ] relevant documentation added or amended (if needed)

### CODE REVIEW

File: kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        setNotifications(prev_notifs => mergeArraysByKey(prev_notifs, current_notifs, 'announceId'));
+      })
+      .catch(error => {
+        console.error('There was a problem checking for Notifications:', error);
+      });
+  };
+
+  useEffect(() => {
+    const announce = window._clientConfig.announcements;
+    const checkUrl = window._clientConfig.telemetryUri;
```
Comment (Conalb97) [inline kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx:157]: check url is assigned a value but never used

**Review by Conalb97 — COMMENTED** [review]
A couple of small comments, but otherwise looks good to me :)

**Review by rebecca-thompson — COMMENTED** [review]
This is looking great! A couple of things I've noticed...

It looks like it's possible to crash the app when a date field is malformed or in an unaccepted format in the config. I'm not sure if that's what you want, or if you'd prefer to defend against and omit announcements not in the correct format.
Screenshot 2024-04-11 at 17 02 43

Less dramatic, but the enum fields `category` and `lifespan` also have some interesting effects in the ui if they are misspelled or have a different value.


A number of functions are declared in the top level of the React component - ideally it would be great to either [cache these functions between re-renders](https://react.dev/reference/react/useCallback), or move them out of the component as utility functions.

Lastly (and this is more of a product question, than a coding question), it felt unexpected that, when clicking on the 'Please click to acknowledge' error bar, that the bar did not disappear, but some of the other messages do. It feels like the whole bar should be clickable to dismiss (rather than just the cross). It also feels like the other message should not disappear when one message is dismissed.

https://github.com/guardian/grid/assets/33927854/a130e154-f051-42ca-8292-54ae9a4d9af9

File: kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    if (notif_cookie) {
+      const current_notif_ids = announce.map(ann => ann.announceId).join(",");
+      const notif_ids = notif_cookie.split(',');
+      const new_notif_ids = notif_ids.filter(n_id => current_notif_ids.includes(n_id)).join(",");
+      document.cookie = (`${NOTIFICATION_COOKIE}=${new_notif_ids}; max-age=${cookie_age}`);
+    }
+
+    // Clean up the event listener when the component unmounts
+    return () => {
+      window.removeEventListener("mouseup", autoHideListener);
```
[the remark this answers is not in the document]
Comment (AndyKilmory) [reply kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx:184]: corrected

File: kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+export interface NotificationBannerProps {
+  heading: string;
+}
+
+const NotificationsBanner: React.FC<NotificationBannerProps> = ({heading}) => {
+  const bannerHeading = heading;
+  const initNotifications: Notification[] = [];
+  const [notifications, setNotifications] = useState(initNotifications);
+  let checkNotificationsRef: NodeJS.Timeout;
```
[the remark this answers is not in the document]
Comment (AndyKilmory) [reply kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx:79]: As the work to set the timer is done within the initialisation useEffect (ie. no dependencies) it only gets called when component initialises not on rerenders - so I think it is perhaps okay? I've tidied the code to embed the full declaration of the timer within the initialisation useEffect to make its operation a clearer/cleaner. It appears that the timer gets correctly initialised on full page refresh and start up - which is ok.

File: kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  urlText: string,
+  category: string,
+  lifespan: string
+}
+
+export interface NotificationBannerProps {
+  heading: string;
+}
+
+const NotificationsBanner: React.FC<NotificationBannerProps> = ({heading}) => {
```
[the remark this answers is not in the document]
Comment (AndyKilmory) [reply kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx:75]: removed

File: kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  const bannerHeading = heading;
+  const initNotifications: Notification[] = [];
+  const [notifications, setNotifications] = useState(initNotifications);
```
[the remark this answers is not in the document]
Comment (AndyKilmory) [reply kahuna/public/js/components/gr-notifications-banner/gr-notifications-banner.tsx:78]: done

**Comment by AndyKilmory** [comment]
> This is looking great! A couple of things I've noticed...
> 
> It looks like it's possible to crash the app when a date field is malformed or in an unaccepted format in the config. I'm not sure if that's what you want, or if you'd prefer to defend against and omit announcements not in the correct format. Screenshot 2024-04-11 at 17 02 43
> 
> Less dramatic, but the enum fields `category` and `lifespan` also have some interesting effects in the ui if they are misspelled or have a different value.
> 
> A number of functions are declared in the top level of the React component - ideally it would be great to either [cache these functions between re-renders](https://react.dev/reference/react/useCallback), or move them out of the component as utility functions.
> 
> Lastly (and this is more of a product question, than a coding question), it felt unexpected that, when clicking on the 'Please click to acknowledge' error bar, that the bar did not disappear, but some of the other messages do. It feels like the whole bar should be clickable to dismiss (rather than just the cross). It also feels like the other message should not disappear when one message is dismissed.
> Screen.Recording.2024-04-11.at.17.14.06.mov

The issues regarding incorrect date, category and lifespan issues have been addressed providing reasonable defaults given the expected nature of announcements declared in the application config (ie feature announcements to be acknowledged and then not seen again).

In terms of the banner behaviour - it is not expected that in real world scenarios all 3 lifespan types will be present at one instant. The transient  type is designed for simple messaging that we are happy for users to ignore if they want (success messages etc), the session and persistent messages require more user attention and the team deliberately decided to opt for the user having to click the 'x' to acknowledge the message - the choice of the 'x' over other button options was discussed but UX requested the 'x' to be consistent with the IntGEL styling patterns used across BBC Apps.

**Review by rebecca-thompson — APPROVED** [review]

**Comment by AndyKilmory** [comment]
Thanks Rebecca

**Comment by AndyKilmory** [comment]
Hi Rebecca, thanks again!

**Comment by AndyKilmory** [comment]
Thanks for approval. All working as expected now

**Pull request merged**